### DISCUSSION TITLE
ML Experiment Tracking and Management

### DISCUSSION THREAD

**Discussion by naga-karthik** [ivadomed/ivadomed#1004] [Ideas]
The aim of this post is to open a stage for discussing various alternatives for ML experiment tracking. 

### Motivation:

As we train more and more DL models, keeping track of the models (i.e. how many filters were used, what learning rate, epochs, etc.) becomes increasingly difficult. The brute-force way is to manually add entries onto a spreadsheet but there are several much better alternatives that make this task a lot easier.

### Trackers Explored:

- [Weights and Biases](https://wandb.ai/site): This was tested (by me) while working on a course project. My overall experience has been good. There are several useful features, for e.g. grouping experiments into "Runs", visualization of all loss curves, logging of hyperparameters and gradients. Most importantly, everything can be found online (on the user's wandb profile), which avoids running it on localhost (as it is done on Tensorboard). 
   - While it is easy to use on a personal basis (an account can be created free of charge), I am unaware of the potential issues it might create when we want to integrate it into the ivadomed system. Particularly, will it be considered as an "enterprise" account, or, are the users who will use ivadomed expected to create wandb accounts as a prerequisite? Need some discussion here.
   - [Here](https://wandb.ai/site/pricing) are the pricing details for wandb 

- [Comet.ml](https://www.comet.ml/site/): has been tried @dyt811 
   -  Comet.ml has apparently good integration with a wide variety of ML run time environments/framewokrs: troch/tensorflow/SKLearn/MLFlow, etc. 
   - Key commands for logging [Comet.ml cheat sheet: supercharge your machine learning experiment management](https://medium.com/comet-ml/comet-ml-cheat-sheet-supercharge-your-machine-learning-experiment-management-7786e08f8e9e) 
   - [Hyperparameters search](https://www.comet.ml/docs/python-sdk/introduction-optimizer/)
      - Also check out Optuna also for optimization
   - Pro: 
      - Off the shelf, centralized web portal to get started with API keys within literally minutes to log metrics/data
      - Good for metrics/parameters/experiments tracking with minimum amount of code. 
      - compatible with multiple experimental tracking solutions
      - has built in optimizer
      - Probably more commercial support if you are on paid tier? 
      - Free for academia affiliated researchers
   - Con: 
      - Expensive if you have to buy it (a few hundred per month per seat), 
      - not open source
      - not locally deployable AFAIK. 

- [MLFlow](https://mlflow.org): has been around for a while too
    - Pro: 
       - Free. Open source
       - Can be used to reproduce/replicate simple experiments (have not tried run on more complicated deep learning)
       - Probably more community support? 
    - Con:
       - Lack an off the shelf centralized SAAS solution. You will have to host your own (e.g. kind of like a tensorboard portal)
       - Slightly less feature than the Comet.ml solution in the track space. Not 100% sure. ? 

### Future:

In my opinion, having an experiment tracker (that can be accessed online without localhost) is definitely a plus point. It turns out [this issue](https://github.com/ivadomed/ivadomed/issues/801) already brought up this topic and some more information on this can be found here. I hope that this post will motivate more discussions and possibly others can chip in with other (newer) experiment trackers that they might have used.

**Comment by dyt811** [comment]
Locked and moved to Issue  #1010
